### DISCUSSION TITLE
Can Mojo 0.6.1 program microcontrollers?

### DISCUSSION THREAD

**Discussion by Eprahim-taha** [modular/modular#1561] [Q&A]
I am very curious to know whether Mojo 0.6.1 is capable of programming microcontrollers or not, and I do not really know the answer. Please, if you know the answer, So tell me Thanks.

**Comment by Sharktheone** [comment]
I'd say no. As from the answer from #1461, there is some work that would need to be done in order it can be used for microcontrollers. 

One problem that I see is that some microcontrollers have different (exotic) instruction sets. The compiler currently has only real support for x86 and arm and those are tied to the operating systems (arm only works with macOS, ...)

This will probably stay like this for a bit, because the priories lie different (e.g Windows support, bug fixes, ...).

TLDR; No, not with 0.6.1

**Reply by Eprahim-taha** [reply]
Thanks for the clarification 😘

**Comment by rarebreed** [comment]
As for the different exotic instruction sets, I'm not sure where MLIR and LLVM would fit into the picture, but I believe they will help.  I'm still not exactly sure what MLIR does, but it appears to be a "compiler framework" that allows hardware vendors to quickly create backends that will take some kind of IR (I presume LLVM IR) and generate the machine code required by the hardware.

This is necessary for all the exotic GPU and other hardware accelerators also, so I imagine it can be applied to microcontrollers as well.  So, I would say that mojo's long-term goals would make it a good language for baremetal hardware

- No Garbage Collector
- Stand alone binaries
- A way to more easily target hardware arches (eg thumbv8m.main-none-eabihf target triple for ARM cortex M type controllers)

I think what mojo would need, and I haven't heard of anything on the roadmap for, is an abstracted out standard library like rust has, so that the "regular" standard library which makes use of OS resources is built on top of a more core fundamental library that doesn't assume built in memory allocators and data types that need it (e.g. any kind of heap allocated structure, including String types).

**Reply by hcye2022** [reply]
Can mojo compile mojo src into LLVM IR that can be re-targeted to MCUs using clang?

**Reply by rarebreed** [reply]
From what I understand, the mojo compiler only uses LLVM IR for final codegen, but there are currently no flags in the mojo cli to emit only the LLVM IR itself as output. However, the MLIR compiler also seems to slightly tweak the LLVM IR output to make it more suitable for mojo.  For example, they disable LLVM IR inlining and loop unrolling, as well as make it make it much more parallelizable (which gives me high hopes that mojo will have far better compile times than rust which is pretty bad).

There's a good presentation about MLIR and mojo.  Specifically, here is where it is mentioned that MLIR will output LLVM IR, albeit with some modifications.
https://youtu.be/SEwTjZvy8vw?t=1586

**Reply by Sharktheone** [reply]
I guess, one other main problem is that the stdlib has some parts that only work on windows and mac. And since you currently can't disable the stdlib, I think it's not going to work. 😢
